lib/rucio/core/authentication.py

```diff
@@ -349,17 +446,856 @@ def get_auth_token_saml(account, saml_nameid, appid, ip=None, session=None):
+    :param session: The database session in use.
+
+    :returns: new DB access token object
+    """
+    if not grant_type:
+        grant_type = "urn:ietf:params:oauth:grant-type:token-exchange"
+    global OIDC_clients
+    try:
+        start = time.time()
+
+        record_counter(counters='IdP_authentication.code_granted')
+        # Validate the subject token
+        validated = validate_auth_token(subject_token)
+        if not validated:
+            raise CannotAuthenticate("The subject token presented for the token exchange is invalid.")
+        # In case user presents itself with any IAM JWT token to Rucio, Rucio will validate it (signature, identity) and ask
+        # for a token exchange with the audience and scope valid for operations with Rucio
+        if scope is None and audience is None:
+            scope = validated['authz_scope']
+            audience = 'rucio'
+            if 'openid' not in scope or 'profile' not in scope:
+                scope = 'openid profile'
+        # Get the token issuer from the token payload
+        issuer = JWT().unpack(subject_token).payload()['iss']
+        OIDC_client = OIDC_clients[issuer]
+        # create a grant instance of the OIDC_client
+        exchange_session_state = rndstr(50)
+        OIDC_client.grant[exchange_session_state] = Grant()
+        OIDC_client.grant[exchange_session_state].grant_expiration_time = time_util.utc_time_sans_frac() + 60
+        OIDC_client.grant[exchange_session_state].code = "exchange_code"
+        resp = AccessTokenResponse()
+        resp["subject_token"] = subject_token
+        OIDC_client.grant[exchange_session_state].tokens.append(Token(resp))
+
+        # assembling parameters to request a new access token
+        args = {"subject_token": subject_token,
+                "scope": scope,
+                "audience": audience,
+                "grant_type": grant_type}
+        # exchange access code for a access token
+        oidc_token_response = OIDC_client.do_any(Message,
+                                                 endpoint=OIDC_client.provider_info["token_endpoint"],
+                                                 state=exchange_session_state,
+                                                 request_args=args,
+                                                 authn_method="client_secret_basic")
+        oidc_tokens = oidc_token_response.json()
+        if isinstance(oidc_tokens['scope'], list):
+            scope = " ".join(oidc_tokens['scope'])
+        else:
+            scope = oidc_tokens['scope']
+        expired_at = datetime.datetime.utcnow() + datetime.timedelta(seconds=oidc_tokens['expires_in'])
+        if 'refresh_token' in oidc_tokens:
+            # save the received access and refresh tokens in the DB
+            if refresh_lifetime:
+                refresh_lifetime = int(refresh_lifetime)
+            else:
+                refresh_lifetime = 96
```

Comment:
hardcoded by design?

Reply:
I see I have it at several places, just to make sure there will always be a default value. I will go through it in case I could put some away to save lines of code, but in general, we wanted to have some default to which to fallback in case used does not configure this value. Currently this is set to 4 days.